src/groups.py

```diff
@@ -0,0 +1,380 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Helper class used to manage interactions with the Ranger API for user/group operations."""
+
+import copy
+import json
+import logging
+
+import requests
+import yaml
+from apache_ranger.exceptions import RangerServiceException
+from ops.charm import CharmBase
+from ops.model import ActiveStatus
+
+from literals import (
+    ADMIN_USER,
+    ENDPOINT_MAPPING,
+    EXPECTED_KEYS,
+    HEADERS,
+    MEMBER_TYPE_MAPPING,
+    RANGER_URL,
+    SYSTEM_GROUPS,
+)
+from utils import (
+    generate_random_string,
+    handle_service_error,
+    log_event_handler,
+    retry,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class RangerGroupManager:
+    """This class handles the communication with the Ranger API."""
+
+    def __init__(self, charm: CharmBase):
+        """Construct RangerGroupManager object.
+
+        Args:
+            charm: the charm for which this relation is provided
+        """
+        self.charm = charm
+
+    @property
+    def _auth(self):
+        """Return the authentication credentials."""
+        admin_password = self.charm.config["ranger-admin-password"]
+        auth = (ADMIN_USER, admin_password)
+        return auth
+
+    @log_event_handler(logger)
+    def _handle_synchronize_file(self, event):
+        """Synchronize the data from the config file with the Ranger API.
+
+        Args:
+            event: The config changed event that triggered the synchronization.
+        """
+        if not self.charm.unit.status == ActiveStatus("Status check: UP"):
+            logger.debug("Status not active, event deferred.")
+            event.defer()
+            return
+
+        if not self.charm.unit.is_leader():
+            return
+
+        config_data = self._read_file()
+        for key in config_data:
+            data = next(iter(config_data.values()))
+            relation_data = copy.deepcopy(data)
```

Comment:
For my information, is `deepcopy` necessary? Who will write to it that it cannot be shared?

Reply:
This is actually not needed anymore with the refactor, it was needed originally because of the nested users being replaced by ids throughout the workflow but I am now using _state to store these. I have removed it.